### DISCUSSION TITLE
How do I use RT-DETR model with COCO Format Data?

### DISCUSSION THREAD

**Discussion by hpmadness** [ultralytics/ultralytics#13524] [Q&A]
I am trying to train a the `rtdetr-l.yaml` model without pretrained weights on my data that is in COCO Format. So far I've done:

```
from ultralytics import RTDETR
model = RTDETR("/rtdetr-l.yaml")
results = model.train(data="data.yaml", epochs=10, imgsz=512, device=0, project=/training_results', name='TEST', batch=1, verbose=True)
```

My `data.yaml` file looks like this:
```
train: train_images_512_02 # directory to train images
val: val_images_512_02  # directory to val images

# COCO json annotation files
train_ann: train_512_02.json
val_ann: val_512_02.json

# Classes
nc: 60
names: ["class0", "class2" ... "class59"]
```

Now when I run the code in the first code block, I get:
```
train: Scanning train_images_512_02.cache... 0 images, 39171 backgrounds, 0 corrupt: 100%|██████████| 39171/39171 [00:00<?, ?it/s]
WARNING ⚠️ No labels found in train_images_512_02.cache, training may not work correctly. See https://docs.ultralytics.com/datasets for dataset formatting guidance.
val: Scanning val_images_512_02.cache... 0 images, 13059 backgrounds, 0 corrupt: 100%|██████████| 13059/13059 [00:00<?, ?it/s]
WARNING ⚠️ No labels found in val_images_512_02.cache, training may not work correctly. See https://docs.ultralytics.com/datasets for dataset formatting guidance.
Plotting labels to /TEST9/labels.jpg... 
```
I'm not sure why it's getting only background images and why the labels are not being found. Help would be much appreciated.

**Comment by glenn-jocher** [comment]
Hello @hpmadness,

Thank you for reaching out! It looks like you're on the right track with setting up your training for the RT-DETR model using COCO format data. However, the issue seems to be with the dataset annotations not being correctly recognized. Let's troubleshoot this step-by-step.

### Verify Dataset Structure and Annotations

1. **Directory Structure**: Ensure that your `train_images_512_02` and `val_images_512_02` directories contain the images and that the paths specified in `data.yaml` are correct.

2. **COCO Annotations**: The `train_512_02.json` and `val_512_02.json` files should follow the COCO annotation format. Ensure that these JSON files are correctly formatted and contain the necessary annotations for your images.

### Example `data.yaml` for COCO Format

Your `data.yaml` looks mostly correct, but let's ensure it aligns perfectly with the expected format:

```yaml
train: train_images_512_02  # directory to train images
val: val_images_512_02  # directory to val images

# COCO json annotation files
train_ann: train_512_02.json
val_ann: val_512_02.json

# Classes
nc: 60
names: ["class0", "class1", ..., "class59"]
```

### Code Example for Training

Your training code seems correct, but let's ensure you are using the latest versions of `torch` and `ultralytics`:

```python
from ultralytics import RTDETR

# Ensure you have the latest versions
!pip install --upgrade torch ultralytics

# Load the model
model = RTDETR("rtdetr-l.yaml")

# Train the model
results = model.train(
    data="data.yaml",
    epochs=10,
    imgsz=512,
    device=0,
    project='/training_results',
    name='TEST',
    batch=1,
    verbose=True
)
```

### Debugging Steps

1. **Check Annotations**: Verify that your COCO JSON files contain the correct annotations. You can use tools like [COCO Annotator](https://github.com/jsbroks/coco-annotator) to visualize and validate your annotations.

2. **Cache Files**: The warning indicates that no labels were found in the cache files. This usually means that the annotations were not correctly parsed. Ensure that the paths and filenames in your JSON files match the images in your directories.

3. **Verbose Output**: Since you have `verbose=True`, check the detailed logs for any additional clues about why the labels are not being found.

### Additional Resources

For more detailed guidance on dataset formatting, please refer to the [Ultralytics Dataset Documentation](https://docs.ultralytics.com/datasets). This resource provides comprehensive information on how to structure your dataset and annotations for successful training.

If the issue persists, please provide a minimum reproducible example, including a small subset of your dataset and annotations, so we can further investigate. You can find more information on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions or need further assistance. Happy training! 🚀

**Comment by hpmadness** [comment]
Thanks for your response @glenn-jocher . Here is the MRE (to the best of my abilities) of the data. Please let me know if you need further information.
```
import json
with open('train_512_02.json') as json_file:
    train_512_02 = json.load(json_file)
```
`train_512_02 .keys()` = `dict_keys(['images', 'annotations', 'categories'])`

Here's a brief part of what each of the keys looks like:
1. `train_512_02['images']` results in:

```
[{'height': 512, 'width': 512, 'id': 1, 'file_name': '1577_0_0_0_512_512.jpg'},
 {'height': 512,
  'width': 512,
  'id': 2,
  'file_name': '1577_0_410_0_922_512.jpg'},
 {'height': 512,
  'width': 512,
  'id': 3,
  'file_name': '1577_0_820_0_1332_512.jpg'},
 {'height': 512,
  'width': 512,
  'id': 4,
  'file_name': '1577_0_1230_0_1742_512.jpg'}, ...]
```

2. `train_512_02['annotations']` results in:
```
[{'iscrowd': 0,
  'image_id': 34,
  'bbox': [130.0, 280.0, 112.0, 99.0],
  'segmentation': [[130, 280, 130, 379, 242, 379, 242, 280]],
  'category_id': 48,
  'id': 1,
  'area': 11088},
 {'iscrowd': 0,
  'image_id': 42,
  'bbox': [454.0, 357.0, 58.0, 63.0],
  'segmentation': [[454, 357, 454, 420, 512, 420, 512, 357]],
  'category_id': 48,
  'id': 2,
  'area': 3654},
 {'iscrowd': 0,
  'image_id': 42,
  'bbox': [382.0, 115.0, 92.0, 82.0],
  'segmentation': [[382, 115, 382, 197, 474, 197, 474, 115]],
  'category_id': 48,
  'id': 3,
  'area': 7544}, ...]
```
3. `train_512_02['categories']` results in:
```
[{'id': 0,
  'name': 'class0',
  'supercategory': ''class0'},
 {'id': 1, 'name': ''class1', 'supercategory': ''class1'},
 {'id': 2, 'name': 'class2', 'supercategory': 'class2'},
 ...
 {'id': 59, 'name': 'class59', 'supercategory': 'class59'}]
```

I have double checked the paths and they are correct. They point to the folder with all the .jpg images (as in, train_images_512_02 contains all my images for training and same for val_images_512_02 for val). Additionally, train_512_02.json is name and path to my train annotations and val_512_02.json is name and path to my val annotations.

Additional messages in the terminal are:
```
WARNING ⚠️ no model scale passed. Assuming scale='l'.
rtdetr-l summary: 673 layers, 32929376 parameters, 32929376 gradients, 108.3 GFLOPs
Ultralytics YOLOv8.2.31 🚀 Python-3.8.16 torch-2.0.1+cu117 CUDA:0 (NVIDIA A100-SXM4-80GB, 81038MiB)
engine/trainer: task=detect, mode=train, model=rtdetr-l.yaml, data=data.yaml, epochs=10, time=None, patience=100, batch=1, imgsz=512, save=True, save_period=-1, cache=False, device=0, workers=8, project=/training_results, name=TEST9, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=training_results/TEST9
WARNING ⚠️ no model scale passed. Assuming scale='l'.
```
Also, does `model = RTDETR("rtdetr-l.yaml")` give me a pretrained model with weights or just an architecture for the models where I can train it from scratch?

Let me know if you need more information. Thanks.

**Reply by glenn-jocher** [reply]
@hpmadness thank you for providing the detailed information and the minimum reproducible example (MRE). It looks like you're on the right track, but there are a few adjustments needed to ensure your dataset is correctly recognized and utilized by the RT-DETR model.

### Key Points to Address:
1. **Data Format**: Ensure your `data.yaml` file correctly references the COCO format annotations.
2. **Model Initialization**: Clarify whether you want to use a pretrained model or train from scratch.

### Adjusting Your `data.yaml` File:
Your `data.yaml` file should look like this to correctly reference the COCO format annotations:

```yaml
train: train_images_512_02 # directory to train images
val: val_images_512_02  # directory to val images

# COCO json annotation files
train_ann: train_512_02.json
val_ann: val_512_02.json

# Classes
nc: 60
names: ["class0", "class1", ..., "class59"]
```

### Model Initialization:
To train the model from scratch, you should initialize the model with the architecture file (`rtdetr-l.yaml`). If you want to use pretrained weights, you should use the pretrained model file (`rtdetr-l.pt`).

### Training from Scratch:
```python
from ultralytics import RTDETR

# Load the model architecture (without pretrained weights)
model = RTDETR("rtdetr-l.yaml")

# Train the model on your custom dataset
results = model.train(data="data.yaml", epochs=10, imgsz=512, device=0, project='/training_results', name='TEST', batch=1, verbose=True)
```

### Using Pretrained Weights:
```python
from ultralytics import RTDETR

# Load the pretrained model
model = RTDETR("rtdetr-l.pt")

# Train the model on your custom dataset
results = model.train(data="data.yaml", epochs=10, imgsz=512, device=0, project='/training_results', name='TEST', batch=1, verbose=True)
```

### Common Issues and Solutions:
1. **No Labels Found**: The warning `No labels found in train_images_512_02.cache` indicates that the annotations are not being correctly read. Ensure that the paths in your `data.yaml` file are correct and that the COCO JSON files are properly formatted.
2. **Model Scale Warning**: The warning `no model scale passed. Assuming scale='l'` is informational and indicates that the default scale 'l' is being used. This is expected behavior unless you specify a different scale.

### Additional Debugging Steps:
1. **Check JSON Annotations**: Ensure that your COCO JSON files (`train_512_02.json` and `val_512_02.json`) are correctly formatted and contain the necessary fields (`images`, `annotations`, `categories`).
2. **Verify Image Paths**: Ensure that the image paths in the JSON files match the actual file paths in your directories.

### Example JSON Structure:
Here’s a brief example of how your COCO JSON should be structured:

```json
{
  "images": [
    {"height": 512, "width": 512, "id": 1, "file_name": "1577_0_0_0_512_512.jpg"},
    ...
  ],
  "annotations": [
    {"iscrowd": 0, "image_id": 1, "bbox": [130.0, 280.0, 112.0, 99.0], "category_id": 48, "id": 1, "area": 11088},
    ...
  ],
  "categories": [
    {"id": 0, "name": "class0"},
    ...
  ]
}
```

### Final Notes:
- Ensure you have the latest versions of `torch` and `ultralytics` installed.
- If the issue persists, please provide any additional error messages or logs.

For more detailed information, you can refer to the [Ultralytics RT-DETR documentation](https://docs.ultralytics.com/models/rtdetr/).

Feel free to reach out if you have any further questions or need additional assistance. Happy training! 🚀

**Reply by hpmadness** [reply]
Hello @glenn-jocher ,

I'm comparing what you wrote for the `data.yaml` file in your most recent response and what I copied and pasted of what my `data.yaml` and I don't see a difference. Am I missing something?

Secondly, thanks for verifying that I'm loading in the model correctly to train it from scratch.

Thirdly, for the JSON structure the things I see different is that your `annotations` don't have segmentation in them. Also the `categories 
 ` key,value pair doesn't have the `'supercategory'` key. Are these the only differences and what causes it to not work?

**Reply by glenn-jocher** [reply]
Hello @hpmadness,

Thank you for your detailed follow-up! Let's address your points one by one:

1. **`data.yaml` File**:
   It looks like your `data.yaml` file is correctly formatted. The key is ensuring that the paths and filenames are accurate and accessible. If everything matches, this part should be fine.

2. **Model Initialization**:
   Great to hear that the model initialization is clear now. Training from scratch with `rtdetr-l.yaml` should work as expected.

3. **JSON Structure**:
   The presence of the `segmentation` key in your annotations and the `supercategory` key in your categories should not cause issues if they are correctly formatted. However, the warning about no labels found suggests there might be an issue with how the annotations are being read.

   Here are a few things to check:
   - **Segmentation**: Ensure that the `segmentation` field is correctly formatted as a list of lists, even if it's empty.
   - **Category IDs**: Verify that the `category_id` in annotations matches the `id` in categories.
   - **Image IDs**: Ensure that the `image_id` in annotations matches the `id` in images.

   Here’s a quick checklist:
   - **Annotations**:
     ```json
     {
       "iscrowd": 0,
       "image_id": 1,
       "bbox": [130.0, 280.0, 112.0, 99.0],
       "segmentation": [[130, 280, 130, 379, 242, 379, 242, 280]],
       "category_id": 48,
       "id": 1,
       "area": 11088
     }
     ```
   - **Categories**:
     ```json
     {
       "id": 0,
       "name": "class0",
       "supercategory": "class0"
     }
     ```

   If these elements are correctly formatted and aligned, the dataset should be read correctly.

### Additional Debugging Steps:
- **Verify Paths**: Double-check that the paths in `data.yaml` correctly point to the directories containing your images and JSON files.
- **Update Packages**: Ensure you are using the latest versions of `torch` and `ultralytics`:
  ```bash
  pip install --upgrade torch ultralytics
  ```

### Example Code:
Here’s a quick example to verify the dataset loading:
```python
from ultralytics import RTDETR

# Load the model architecture (without pretrained weights)
model = RTDETR("rtdetr-l.yaml")

# Train the model on your custom dataset
results = model.train(data="data.yaml", epochs=10, imgsz=512, device=0, project='/training_results', name='TEST', batch=1, verbose=True)
```

If the issue persists, please provide any additional error messages or logs. This will help us diagnose the problem more effectively.

Feel free to reach out if you have any further questions or need additional assistance. Happy training! 🚀
